pxd_fastpath.c

```diff
@@ -1198,18 +906,52 @@ static void pxd_suspend_io(struct pxd_device *pxd_dev)
 
 static void pxd_resume_io(struct pxd_device *pxd_dev)
 {
+	LIST_HEAD(tmpQ);
 	bool wakeup;
-	spin_lock(&pxd_dev->fp.suspend_wait.lock);
-	pxd_dev->fp.suspend--;
-	wakeup = (pxd_dev->fp.suspend == 0);
-	spin_unlock(&pxd_dev->fp.suspend_wait.lock);
-
+	int cpu, new = 0, old = 0;
+
+	BUG_ON(!pxd_dev->fp.state);
+	spin_lock(&pxd_dev->fp.suspend_lock);
+	for_each_online_cpu(cpu) {
```

Comment:
If both susp/resume and going to iterate over all cpus, it needs locked lookup. It also means local cpu reads need lock. Why use percpu at all?
If u are going to use percpu, doesnt it make more sense to partition the fp.suspend_queue, so list_add is done percpu?

Reply:
Only the control operation (i.e. a suspend or a resume call) needs to ensure all cpu have the same state. 
Regular IO path through make_request() just checks current state only within its local cpu ptr.